sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp

```diff
@@ -967,23 +1042,48 @@ lsc_block_store(AccessorTy acc, uint32_t offset,
   lsc_block_store<T, NElts, DS, L1H>(
       __ESIMD_DNS::accessorToPointer<T>(acc, offset), vals, pred);
 #else
-  detail::check_lsc_vector_size<NElts>();
   detail::check_lsc_data_size<T, DS>();
-  detail::check_lsc_cache_hint<detail::lsc_action::store, L1H, L3H>();
+  detail::check_lsc_cache_hint<detail::lsc_action::load, L1H, L3H>();
   constexpr uint16_t _AddressScale = 1;
   constexpr int _ImmOffset = 0;
   constexpr lsc_data_size _DS = detail::finalize_data_size<T, DS>();
-  static_assert(_DS == lsc_data_size::u32 || _DS == lsc_data_size::u64,
-                "Transposed store is supported only for data size u32 or u64");
-  constexpr detail::lsc_vector_size _VS = detail::to_lsc_vector_size<NElts>();
   constexpr detail::lsc_data_order _Transposed =
       detail::lsc_data_order::transpose;
   constexpr int N = 1;
+
   __ESIMD_NS::simd<uint32_t, N> offsets = offset;
   auto si = __ESIMD_NS::get_surface_index(acc);
-  __esimd_lsc_store_bti<T, L1H, L3H, _AddressScale, _ImmOffset, _DS, _VS,
-                        _Transposed, N>(pred.data(), offsets.data(),
-                                        vals.data(), si);
+
+  if constexpr (_DS == lsc_data_size::u32 || _DS == lsc_data_size::u64) {
+    detail::check_lsc_vector_size<NElts>();
+    constexpr detail::lsc_vector_size _VS = detail::to_lsc_vector_size<NElts>();
+    __esimd_lsc_store_bti<T, L1H, L3H, _AddressScale, _ImmOffset, _DS, _VS,
+                          _Transposed, N>(pred.data(), offsets.data(),
+                                          vals.data(), si);
+  } else if constexpr (_DS == lsc_data_size::u16) {
+    static_assert(NElts % 2 == 0,
+                  "Number of elements is not supported by Transposed store");
+    detail::check_lsc_vector_size<NElts / 2>();
```

Comment:
If we choose to translate loads/stores of 8 and 16-bit elements to loads/stores of 32-bit elements that allows load/store
MaxN*2 16-bit elements and MaxN*4 8-bit elements.
Don't we want to use 64-bit loads/stores when it is possible? In those cases it would allow loading/storing 2x longer memory chunk of 8/16-bit elements, i.e. MaxN*4 for 16-bit elements and MaxN*8 for 8-bit elements.

Reply:
Good point, @v-klochkov. We need to understand what are requirements for the hardware instruction first, and add them to the doxygen comments of the function as a requirement. Similarly to https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/intel/esimd/memory.hpp#L272.

Note that for block_load we also have alignment template parameter, depending on which one or the other intrinsic is generated. Here I think we always generate single intrinsic and it does not allow to convey alignment info, right? Otherwise we might need to add alignment parameter here as well.

Reply:
the address passed to intrinsic (and HW instruction) must be 4-byte aligned if the load/store of 32-bit elements, and 8-byte aligned for 64-bit element vectors loads/stores.

Reply:
@fineg74 can you please add a comment explaining the requirement to alignment of the address passed to the intrinsics.